src/main/java/appeng/parts/p2p/PartP2PFEPower.java

```diff
@@ -124,129 +149,57 @@ public int receiveEnergy( int maxReceive, boolean simulate )
 			return total;
 		}
 
-		return 0;
-	}
+		@Override
+		public boolean canExtract()
+		{
+			return false;
+		}
 
-	@Override
-	public int extractEnergy( int maxExtract, boolean simulate )
-	{
-		return 0;
-	}
+		@Override
+		public boolean canReceive()
+		{
+			return true;
+		}
 
-	private IEnergyStorage getOutput()
-	{
-		if( this.isOutput() )
+		@Override
+		public int getMaxEnergyStored()
 		{
-			if( !this.cachedTarget )
-			{
-				final TileEntity self = this.getTile();
-				final TileEntity te = self.getWorld().getTileEntity( new BlockPos( self.getPos().getX() + this.getSide().xOffset, self.getPos()
-						.getY() + this.getSide().yOffset, self.getPos().getZ() + this.getSide().zOffset ) );
+			int total = 0;
 
-				if( te != null && te.hasCapability( Capabilities.FORGE_ENERGY, this.getSide().getOpposite().getFacing() ) )
-				{
-					this.outputTarget = te.getCapability( Capabilities.FORGE_ENERGY, this.getSide().getOpposite().getFacing() );
-				}
-				else
+			try
+			{
+				for( PartP2PFEPower t : PartP2PFEPower.this.getOutputs() )
 				{
-					this.outputTarget = null;
+					total += t.getOutput().getMaxEnergyStored();
 				}
-
-				this.cachedTarget = true;
 			}
-
-			if( this.outputTarget != null && this.outputTarget.canReceive() )
+			catch( GridAccessException e )
 			{
-				return this.outputTarget;
+				return 0;
 			}
-		}
 
-		return NULL_ENERGY_STORAGE;
-	}
-
-	@Override
-	public int getEnergyStored()
-	{
-		if( this.isOutput() || !this.isActive() )
-		{
-			return 0;
+			return total;
 		}
 
-		int total = 0;
-
-		try
+		@Override
+		public int getEnergyStored()
 		{
-			for( PartP2PFEPower t : this.getOutputs() )
+			int total = 0;
+
+			try
 			{
-				total += t.getOutput().getEnergyStored();
+				for( PartP2PFEPower t : PartP2PFEPower.this.getOutputs() )
+				{
+					total += t.getOutput().getEnergyStored();
+				}
 			}
-		}
-		catch( GridAccessException e )
-		{
-			return 0;
-		}
-
-		return total;
-	}
-
-	@Override
-	public int getMaxEnergyStored()
-	{
-		if( this.isOutput() || !this.isActive() )
-		{
-			return 0;
-		}
-
-		int total = 0;
-
-		try
-		{
-			for( PartP2PFEPower t : this.getOutputs() )
+			catch( GridAccessException e )
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS1166)